### DISCUSSION TITLE
How do I change "include-in-header" with light/dark mode?

### DISCUSSION THREAD

**Discussion by jack-davison** [quarto-dev/quarto-cli#6110] [Q&A]
### Description

We use quarto for a lot of our dynamic reporting, and I'm keen to create a dark mode to compliment our light theme.

Part of our template uses `include-in-header` to draw our logo and a simple band across the top of the document. The HTML looks like this: 

```html
<style>
</style>
<body>
<img src="./lightlogo.png" style="float: left; min-width: 45px; width: 10%; max-width: 95px; margin-left: 1%; margin-bottom: 0.5em; margin-top: 0.5em; border-style: none">
<div style="width: 100%; height: 5px; background: #0E4496; overflow: hidden;"></div>
<br>
</body>
```

This creates a document the header of which looks like this:

![image](https://github.com/quarto-dev/quarto-cli/assets/45171616/56bdc42b-6611-4e36-ace5-050c44b8a9eb)

On activating dark mode, I'd like to be able to change the logo to a different version (we have a completely white one) and the band to be white.

However, I can't seem to work out how best to do this!

I'd be open to an alternative that achieves the same look without using `include-in-header` if that simply cannot respond to the light-dark toggle.

(Separately, do we have any control over the placement of the toggle switch, or is it forced to be in the top-right?)

Thanks very much.

**Comment by mcanouil** [comment]
One way is by setting the names of two figures using some defined pattern and to use JavaScript as I describe in <https://mickael.canouil.fr/posts/2023-05-30-quarto-light-dark/>.

**Reply by jack-davison** [reply]
Thanks! I've created a dummy project (https://github.com/jack-davison/quarto-darkmode-temp) and this seems to work, although it also seems to break if you set the document to "self-contained". Is that expected?

**Reply by mcanouil** [reply]
`self-contained` is deprecated as mentioned in Pandoc and Quarto documentations, use `embed-resources` instead to avoir any breakage in the future.
I believe the issue is that Pandoc can't see the resource itself, thus does not embed it.
I'll have to look into this as I did not test it in simple documents (which are likely to be self contained).

**Reply by jack-davison** [reply]
> `self-contained` is deprecated as mentioned in Pandoc and Quarto documentations, use `embed-resources` instead to avoid any breakage in the future

Yes, quite right - force of habit! Though it does also break with `embed-resources`.

Thanks for looking into this - it'd be useful to get something working for simple docs.

I'll not mark anything as the answer for the time being.

**Reply by mcanouil** [reply]
> Yes, quite right - force of habit! Though it does also break with `embed-resources`.

Indeed, because Quarto team kept the former version as an alias when Pandoc deprecated/removed it.
But, in the long range, there is no reason to keep `self-contained`.

**Reply by mcanouil** [reply]
Quick workaround that should work:

````qmd
---
title: "Test Doc"
format: 
  html:
    embed-resources: true
include-in-header: header.html
include-after-body:
  text: |
    <script type="application/javascript">
    // Author: Mickaël Canouil
    // Version: <1.0.0>
    // Description: Change image src depending on body class (quarto-light or quarto-dark)
    // License: MIT
    function updateImageSrc() {
      var bodyClass = window.document.body.classList;
      var images = window.document.getElementsByTagName('img');
      for (var i = 0; i < images.length; i++) {
        var image = images[i];
        var src = image.src;
        var newSrc = src;
        if (bodyClass.contains('quarto-light') && src.includes('.dark')) {
          newSrc = src.replace('.dark', '.light');
        } else if (bodyClass.contains('quarto-dark') && src.includes('.light')) {
          newSrc = src.replace('.light', '.dark');
        }
        if (newSrc !== src) {
          image.src = newSrc;
        }
      }
    }

    var observer = new MutationObserver(function(mutations) {
      mutations.forEach(function(mutation) {
        if (mutation.type === 'attributes' && mutation.attributeName === 'class') {
          updateImageSrc();
        }
      });
    });

    observer.observe(window.document.body, {
      attributes: true
    });

    updateImageSrc();
    </script>
theme:
  light: [flatly, style.scss]
  dark: [darkly, styledark.scss]
---

Text.
````

**Reply by jack-davison** [reply]
I've added that to my dummy repo <https://github.com/jack-davison/quarto-darkmode-temp> but it doesn't seem to have done anything; the logo still doesn't toggle when `embed-resources: true`

``` r
quarto::quarto_version()
#> [1] '1.3.361'
```

<sup>Created on 2023-07-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by mcanouil** [reply]
Ok, I'll check this out later.
Thanks for the feedback.

**Reply by mcanouil** [reply]
It does not work in embed resources because images are no longer files, they are encoded as base64 image, thus, the selector in the JavaScript is no longer valid.
If you include the image as HTML, you can tweak the JavaScript to use CSS class, ID, etc. as the selector.

**Reply by jack-davison** [reply]
Thanks very much for looking into this for me, though could you please elaborate on what you mean by including the images as HTML? My HTML skills are a bit lacking - they do the job but I'm no expert!

**Reply by mcanouil** [reply]
Currently in your example, you include your logo using HTML.
You can add a special id or class, that you target in the JavaScript script.

**Comment by baptiste** [comment]
Just as an alternative, you might consider using SVG for your logo, which could then [inherit colors from the parent document via CSS](https://melanie-richards.com/blog/currentcolor-svg-hcm/)
